near-sdk/src/store/unordered_set/mod.rs

```diff
@@ -0,0 +1,798 @@
+mod impls;
+mod iter;
+
+use super::{FreeList, LookupMap, ERR_INCONSISTENT_STATE};
+use crate::crypto_hash::{CryptoHasher, Sha256};
+use crate::store::free_list::FreeListIndex;
+use crate::store::unordered_set::iter::{
+    Difference, Drain, Intersection, Iter, SymmetricDifference, Union,
+};
+use crate::{env, IntoStorageKey};
+use borsh::{BorshDeserialize, BorshSerialize};
+use std::borrow::Borrow;
+use std::fmt;
+
+pub struct UnorderedSet<T, H = Sha256>
+where
+    T: BorshSerialize + Ord,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    elements: FreeList<T>,
+    index: LookupMap<T, FreeListIndex, H>,
+}
+
+//? Manual implementations needed only because borsh derive is leaking field types
+// https://github.com/near/borsh-rs/issues/41
+impl<T, H> BorshSerialize for UnorderedSet<T, H>
+where
+    T: BorshSerialize + Ord,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    fn serialize<W: borsh::maybestd::io::Write>(
+        &self,
+        writer: &mut W,
+    ) -> Result<(), borsh::maybestd::io::Error> {
+        BorshSerialize::serialize(&self.elements, writer)?;
+        BorshSerialize::serialize(&self.index, writer)?;
+        Ok(())
+    }
+}
+
+impl<T, H> BorshDeserialize for UnorderedSet<T, H>
+where
+    T: BorshSerialize + Ord,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, borsh::maybestd::io::Error> {
+        Ok(Self {
+            elements: BorshDeserialize::deserialize(buf)?,
+            index: BorshDeserialize::deserialize(buf)?,
+        })
+    }
+}
```

Comment:
```suggestion
#[derive(BorshDeserialize, BorshSerialize)]
pub struct UnorderedSet<T, H = Sha256>
where
    T: BorshSerialize + Ord,
    H: CryptoHasher<Digest = [u8; 32]>,
{
    elements: FreeList<T>,
    index: LookupMap<T, FreeListIndex, H>,
}
```

There are no private types leaked here, can just derive